odo/backends/sql.py

```diff
@@ -794,13 +794,14 @@ def select_or_selectable_to_frame(el, bind=None, dshape=None, **kwargs):
     for field, dtype in fields:
         if isinstance(dtype, Option):
             ty = dtype.ty
-            if ty in datashape.integral:
-                dtypes[field] = 'float64'
+            try:
+                dtypes[field] = ty.to_numpy_dtype()
+            except TypeError:
+                dtypes[field] = np.dtype(object)
             else:
-                try:
-                    dtypes[field] = ty.to_numpy_dtype()
-                except TypeError:
-                    dtypes[field] = np.dtype(object)
+                if np.issubdtype(dtypes[field], np.integer):
```

Comment:
Does this else-if branch leave the possibility of no type being converted for that field?

Reply:
It's a try/except/else branch so the `if np.issubdtype(dtypes[field], np.integer)` block runs only if the try has already *succeeded*:
```python
try:
    dtypes[field] = ty.to_numpy_dtype()
except TypeError:
    dtypes[field] = np.dtype(object)
else:
    # If `ty.to_numpy_dtype()` results in an integer type override `dtypes[field]` to be float64
    # We do this because we already know `dtype` is an `Option` type and pandas uses NaN
    # as a null placeholder requiring that the type be floating point rather than integral
    if np.issubdtype(dtypes[field], np.integer):
        # cast nullable ints to float64 so NaN can be used for nulls
        dtypes[field] = np.float64
```

Reply:
26e300e5219cbc667831b54ba55254a58fae75ba is a self-contained repro of the underlying issue